Text is a string data type. Any data type written as text is a string. Any data under single, double or triple quote are strings. There are different string methods and built-in functions to deal with string data types. To check the length of a string use the len() method.

### Creating a String
Strings can be created by enclosing characters inside a single quote or double-quotes. Even triple quotes can be used in Python but generally used to represent multiline strings and docstrings.

In [ ]:
letter = 'P'                # A string could be a single character or a bunch of texts
print(letter)               # P
print(len(letter))          # 1
greeting = 'Hello, World!'  # String could be made using a single or double quote,"Hello, World!"
print(greeting)             # Hello, World!
print(len(greeting))        # 13
sentence = "I hope you are enjoying 30 days of Python Challenge"
print(sentence)

Multiline string is created by using triple single (''') or triple double quotes ("""). See the example below.

In [ ]:
multiline_string = '''I am a teacher and enjoy teaching.
I didn't find anything as rewarding as empowering people.
That is why I created 30 days of python.'''
print(multiline_string)

# Another way of doing the same thing
multiline_string = """I am a teacher and enjoy teaching.
I didn't find anything as rewarding as empowering people.
That is why I created 30 days of python."""
print(multiline_string)

**String Concatenation**  


We can connect strings together. Merging or connecting strings is called concatenation. See the example below:

In [ ]:
first_name = 'Asabeneh'
last_name = 'Yetayeh'
space = ' '
full_name = first_name  +  space + last_name
print(full_name) # Asabeneh Yetayeh
# Checking the length of a string using len() built-in function
print(len(first_name))  # 8
print(len(last_name))   # 7
print(len(first_name) > len(last_name)) # True
print(len(full_name)) # 16

**Escape Sequences in Strings**

In Python and other programming languages \ followed by a character is an escape sequence. Let us see the most common escape characters:

- \n: new line
- \t: Tab means(8 spaces)
- \\: Back slash
- \': Single quote (')
- \": Double quote (")  
Now, let us see the use of the above escape sequences with examples.

In [4]:
print('I hope everyone is enjoying the Python Challenge.\nAre you ?') # line break
print('Days\tTopics\tExercises') # adding tab space or 4 spaces 
print('Day 1\t5\t5')
print('Day 2\t6\t20')
print('Day 3\t5\t23')
print('Day 4\t1\t35')
print('This is a backslash  symbol (\\)') # To write a backslash
print('In every programming language it starts with \"Hello, World!\"') # to write a double quote inside a single quote

I hope everyone is enjoying the Python Challenge.
Are you ?
Days	Topics	Exercises
Day 1	5	5
Day 2	6	20
Day 3	5	23
Day 4	1	35
This is a backslash  symbol (\)
In every programming language it starts with "Hello, World!"


**String formatting / String Interpolation / f-Strings**

In [5]:
a = 4
b = 3
print(f'{a} + {b} = {a +b}')
print(f'{a} - {b} = {a - b}')
print(f'{a} * {b} = {a * b}')
print(f'{a} / {b} = {a / b:.2f}')
print(f'{a} % {b} = {a % b}')
print(f'{a} // {b} = {a // b}')
print(f'{a} ** {b} = {a ** b}')

4 + 3 = 7
4 - 3 = 1
4 * 3 = 12
4 / 3 = 1.33
4 % 3 = 1
4 // 3 = 1
4 ** 3 = 64


**Slicing Python Strings**


In [ ]:
language = 'Python'
first_three = language[0:3] # starts at zero index and up to 3 but not include 3
print(first_three) #Pyt
last_three = language[3:6]
print(last_three) # hon
# Another way
last_three = language[-3:]
print(last_three)   # hon
last_three = language[3:]
print(last_three)   # hon

**Reversing a String**

In [ ]:
greeting = 'Hello, World!'
print(greeting[::-1]) # !dlroW ,olleH

## String Methods
https://www.w3schools.com/python/python_ref_string.asp

In [1]:
dir(str)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'stri

![alt text](images/string-methods.png)

In [6]:
s = 'amir afshari'

In [7]:
s.capitalize() 

'Amir afshari'

In [8]:
s.title()

'Amir Afshari'

**Excercises**
================



## Concatenation

1. Concatenate the string 'Thirty', 'Days', 'Of', 'Python' to a single string, 'Thirty Days Of Python'.
2. Concatenate the string 'Coding', 'For', 'All' to a single string, 'Coding For All'.
3. Declare a variable named company and assign it to an initial value "Coding For All".
4. Print the variable company using `print()`.



## String Manipulation


1. Print the length of the company string using `len()` method and print().
2. Change all the characters to uppercase letters using `upper()` method.
3. Change all the characters to lowercase letters using `lower()` method.
4. Use `capitalize()`, `title()`, `swapcase()` methods to format the value of the string Coding For All.
5. Cut(slice) out the first word of Coding For All string.
6. Check if Coding For All string contains a word Coding using the method `index`, `find` or other methods.
7. Replace the word coding in the string 'Coding For All' to Python.
8. Change Python for Everyone to Python for All using the replace method or other methods.
9. Split the string 'Coding For All' using space as the separator (`split()`).
10. Split the string "Facebook, Google, Microsoft, Apple, IBM, Oracle, Amazon" at the comma.

## Indexing and Slicing


1. What is the character at index 0 in the string Coding For All.
2. What is the last index of the string Coding For All.
3. What character is at index 10 in "Coding For All" string.

## Acronyms and Abbreviations

1. Create an acronym or an abbreviation for the name 'Python For Everyone'.
2. Create an acronym or an abbreviation for the name 'Coding For All'.

## Indexing and Searching


1. Use index to determine the position of the first occurrence of C in Coding For All.
2. Use index to determine the position of the first occurrence of F in Coding For All.
3. Use `rfind` to determine the position of the last occurrence of l in Coding For All People.
4. Use index or `find` to find the position of the first occurrence of the word 'because' in the following sentence: 'You cannot end a sentence with because because because is a conjunction'
5. Use `rindex` to find the position of the last occurrence of the word because in the following sentence: 'You cannot end a sentence with because because because is a conjunction'
6. Slice out the phrase 'because because because' in the following sentence: 'You cannot end a sentence with because because because is a conjunction'









## String Methods


1. Does ''Coding For All' start with a substring Coding?
2. Does 'Coding For All' end with a substring coding?
3. '   Coding For All      '  , remove the left and right trailing spaces in the given string.

## Identifiers

1. Which one of the following variables return True when we use the method `isidentifier()`:
	+ 30DaysOfPython
	+ thirty_days_of_python

## String Formatting


1. Join the list with a hash with space string: ['Django', 'Flask', 'Bottle', 'Pyramid', 'Falcon']
2. Use the new line escape sequence to separate the following sentences.
3. Use a tab escape sequence to write the following lines.
4. Use string formatting method to display the following:
	+ radius = 10
	+ area = 3.14 * radius ** 2
	+ The area of a circle with radius 10 is 314 meters square.
5. Make the following using string formatting methods:
	+ 8 + 6 = 14
	+ 8 - 6 = 2
	+ 8 * 6 = 48
	+ 8 / 6 = 1.33
	+ 8 % 6 = 2
	+ 8 // 6 = 1
	+ 8 ** 6 = 262144